### ISSUE TITLE
vLLM Integration

### ISSUE THREAD

**Issue by robertgshaw2-neuralmagic** [pytorch-labs/applied-ai#17]
FYI - ongoing PR to integrate this great work into vLLM (https://github.com/vllm-project/vllm/pull/3905)

I ran into a couple correctness issues with a few shapes running through our CI

```bash
git clone https://github.com/neuralmagic/nm-vllm.git
cd vllm
git checkout fused-moe
pip install -e .
pip install requirements-dev.txt
pytest -v  tests/kernels/test_moe.py::test_fused_moe
```

**Comment by robertgshaw2-neuralmagic** [comment]
Note: I am able to get all tests passing in vllm if I only run the following shapes:

```python
@pytest.mark.parametrize("m", [2, 4, 8, 16, 32, 64, 128, 512, 1024, 2048])
@pytest.mark.parametrize("n", [14336//2])
@pytest.mark.parametrize("k", [4096])
@pytest.mark.parametrize("e", [8])
@pytest.mark.parametrize("topk", [2])
@pytest.mark.parametrize("dtype", [torch.float16])
```

This matches the test suite of this repo 
- https://github.com/pytorch-labs/applied-ai/blob/main/kernels/triton/inference/col_major_moe_gemm/test_moe_gemm.py#L33


The test suite we have in vllm runs the following:

```python
@pytest.mark.parametrize("m", [512, 222, 33, 1])
@pytest.mark.parametrize("n", [2048, 256, 1024])
@pytest.mark.parametrize("k", [128, 511, 1024])
@pytest.mark.parametrize("e", [8, 64])
@pytest.mark.parametrize("topk", [2, 6])
@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
```

**Comment by AdnanHoque** [comment]
Hi @robertgshaw2-neuralmagic, thanks for running these! Can you try the V1 kernel we have in the meantime, while we debug this?

The V1 kernel features the SplitK Optimization: [https://github.com/pytorch-labs/applied-ai/blob/main/kernels/triton/inference/col_major_moe_gemm/v1_moe_fused.py](https://github.com/pytorch-labs/applied-ai/blob/main/kernels/triton/inference/col_major_moe_gemm/v1_moe_fused.py)

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks for taking a look to fix them. Let me know where you end up

to be clear the v1 kernels I should see ~10-20% speedup right?

**Comment by AdnanHoque** [comment]
Indeed! 18% on A100 and 20% on H100 is what we measured, with the parameters already present in V1.

You might be able to squeeze out more if  a "split_k" parameter is added here: [https://github.com/vllm-project/vllm/blob/f46864d68dfb46ff88f574e6844f10fdb14cd3b5/benchmarks/kernels/benchmark_mixtral_moe.py#L24](https://github.com/vllm-project/vllm/blob/f46864d68dfb46ff88f574e6844f10fdb14cd3b5/benchmarks/kernels/benchmark_mixtral_moe.py#L24)

That would run through a SplitK setting of 2 - 32, and pick the best SplitK and block size parameters based on the problem size and GPU spec.

**Comment by robertgshaw2-neuralmagic** [comment]
Does it ever make sense to use the SplitK over the ColumnMajor?

**Comment by AdnanHoque** [comment]
Short Answer:
No, not if we are able to figure out the ColumnMajor fix :) 

Long Answer:
Ideally we can use both in tandem, as they are both "inference" optimizations. However, adding in ColumnMajor + SplitK proved to be a little more tricky. There may be a GEMM schedule that is extremely well optimized for a SplitK work decomposition that has the column major properties i.e.  maximum reuse of the activation matrix, while also optimally saturating the SMs of the GPU when M < N, K.

**Comment by robertgshaw2-neuralmagic** [comment]
I see, we can add an oracle for instance if SplitK is better in a low M regime

**Comment by robertgshaw2-neuralmagic** [comment]
Anyways thanks for your great work! Ill check back in in a day or so to see how progress goes

**Comment by robertgshaw2-neuralmagic** [comment]
Indeed making the change suggested here https://github.com/pytorch-labs/applied-ai/issues/16 gets the tests to pass

**Comment by AdnanHoque** [comment]
Yeah -- it appears that there is a bug with the column-major code and with the fix it appears the speedup from this optimization is gone, which is quite unfortunate.

In this case we can turn to SplitK (V1) (although bfloat16 is not supported as of yet for tl.atomic_add) as I think you'll be able to see some gain from it. Looks like it was able to provide some speedup on the FP8 side of things: [https://github.com/vllm-project/vllm/pull/3954](https://github.com/vllm-project/vllm/pull/3954)

**Comment by robertgshaw2-neuralmagic** [comment]
That's too bad!

I also noticed that the results are very close in the cases I created as well).

**Comment by chuangzhidan** [comment]
> Indeed! 18% on A100 and 20% on H100 is what we measured, with the parameters already present in V1.
> 
> You might be able to squeeze out more if a "split_k" parameter is added here: https://github.com/vllm-project/vllm/blob/f46864d68dfb46ff88f574e6844f10fdb14cd3b5/benchmarks/kernels/benchmark_mixtral_moe.py#L24
> 
> That would run through a SplitK setting of 2 - 32, and pick the best SplitK and block size parameters based on the problem size and GPU spec.

how to get this 18% -20% boost in vllm ,no need to be done because it's a default setting?
